### PR TITLE
Autocloseable tasks and reports storing

### PR DESCRIPTION
**Pull request by mrgambal**
- fixed bug in AbstractTaskType.get_next. Also we now look only at AbstractTask.closed flag
- user update moved from app.py to AbstractCloseableTask.save_task_result
- autoclose logics has been moved to particular method which should be overridden (example in DummyTaskType)
- restored basic answers

TODO:
- wrap it using our own exceptions;
- some stats;

### CODE REVIEW

File: vulyk/models/task_types.py [hunk 1]
```diff
@@ -1,21 +1,35 @@
...
+    type_name = ""
+
+    redundancy = 3
 
     def __init__(self, redundancy):
         self.redundancy = redundancy
 
         assert issubclass(self.task_model, AbstractTask), \
             "You should define task_model property"
+        # I'm too exhausted to understand why it doesn't work =(
```
Comment (dchaplinsky) [inline vulyk/models/task_types.py]: Still exhausted?

File: vulyk/models/task_types.py [hunk 2]
```diff
@@ -63,78 +77,201 @@ def export_reports(self, qs=None):
...
             task_type=self.type_name,
-            users_count__lt=self.redundancy,
             users_processed__ne=user.id,
-            users_skipped__ne=user.id)
-        res = res[random.randint(0, res.count())]
+            users_skipped__ne=user.id,
+            closed__ne=True)
+
+        if rs.count() > 0:
+            task = rs[random.randint(0, rs.count())]
```
Comment (dchaplinsky) [inline vulyk/models/task_types.py]: else?

Comment (mrgambal) [inline vulyk/models/task_types.py]: else - task = None => return task

File: vulyk/models/task_types.py [hunk 3]
```diff
@@ -63,78 +77,201 @@ def export_reports(self, qs=None):
...
+        :param user: an instance of User model who provided an answer
+        :type user: models.User
+        """
+        task.users_count += 1
+        task.users_processed.append(user)
+
+        if self._is_ready_for_autoclose(answer, task):
+            task.closed = True
+        else:
+            task.closed = task.users_count == self.redundancy
```
Comment (dchaplinsky) [inline vulyk/models/task_types.py]: :+1:

File: vulyk/models/task_types.py [hunk 4]
```diff
@@ -63,78 +77,201 @@ def export_reports(self, qs=None):
...
+        """
+        # TODO: store id of active session in cookies or elsewhere
+        rs = WorkSession.objects(user=user, task=task).order_by('-start_time')
+
+        if rs.count() > 0:
+            rs.first() \
+                .update(
+                set__end_time=datetime.now(),
+                set__answer=answer,
+                set__corrections=answer.corrections)
```
Comment (dchaplinsky) [inline vulyk/models/task_types.py]: Corrections?

Comment (mrgambal) [inline vulyk/models/task_types.py]: Whole amount of actions/corrections given by user in this particular answer. could be used for stats and/or empty reports removal.

File: vulyk/models/task_types.py (hunk 2 above)
Comment (dchaplinsky) [inline vulyk/models/task_types.py]: None, not empty list?

// just asking

Comment (mrgambal) [inline vulyk/models/task_types.py]: Notice: `rs[random.randint(0, rs.count())] ` returns particular task not a QuerySet.
Turnin' back to your Q: we still need the task to be returned due to `WorkSession` creation need.
Look at  line 92.

**Comment by dchaplinsky** [comment]
Ok, here is my 2 cents:
* Current implementation of next endpoint. Aren't they supposed to return an json? Why we are rendering something there?
* Similar question to save_answer endpoint — why form? Don't hesitate to send JSON.
* I think corrections is totally redundant at this stage. Wasn't included to any specs, doesn't give us any advantages, etc.
* Stats/working sessions aren't mandatory as well, but if you want to leave them in the code — leave them.

I think we should focus rather than trying to cover all user cases and mightbes in the very first version. While what we really need is a possibility to show users bloody tasks :)

File: vulyk/models/task_types.py (hunk 2 above)
Comment (dchaplinsky) [inline vulyk/models/task_types.py]: https://docs.python.org/2/library/random.html#random.choice probably? Or QS wouldn't work as sequence here?

On line 92 — I agree, just not sure, what mongoengine will return if there is nothing to return.

**Comment by mrgambal** [comment]
Yep. but we still need to have well-defined:

- installation and extending
- module structure
- assets
- js-modules

File: vulyk/models/task_types.py (hunk 2 above)
Comment (mrgambal) [inline vulyk/models/task_types.py]: hmm. I believe that it uses the approach which is the same or very close to what I've made up in there...

Comment (dchaplinsky) [inline vulyk/models/task_types.py]: Yes, just, you know, is shorter.

**Comment by dchaplinsky** [comment]
Agree, however, corrections, sessions etc wasn't WELL-DEFINED :)
they've wasn't defined at all, to be precise!

Anyway, for points you've mentioned, I think we should just start and see how the code will flow. Hopefully I'll make some progress tonight.

**Comment by mrgambal** [comment]
re:Stats
https://github.com/mrgambal/vulyk/pull/7#commitcomment-9509631

Nevertheless, I'll have fixed this rgd your comments when I get some spare time tomorrow.

**Comment by mrgambal** [comment]
And here come 2 more commits. You're welcome to review.

File: vulyk/app.py [hunk 5]
```diff
@@ -18,33 +20,27 @@
...
+@app.route('/types', methods=['GET'])
 def types():
     """
-    Produces a selectable list of available tasks which are appropriate
+    Produces a list of available tasks types which are appropriate
     for current user.
     """
-    # TODO: need a template
-    return render_template('types.html', types=TASKS_TYPES.keys())
+    res = [t for t in TASKS_TYPES.keys() if g.user.is_eligible_for(t)]
```
Comment (dchaplinsky) [inline vulyk/app.py]: :+1:

**Comment by dchaplinsky** [comment]
We've used to call it beta, because it's beta than nothing.

**Pull request merged**